Fix #563: How to keep data and cache on external drive

dashohoxha opened the pull request:
Fix #563
Replaces https://github.com/iterative/dvc.org/pull/565

jorgeorpinel commented on src/Documentation/sidebar.json:
```diff
@@ -132,6 +132,17 @@
+        "label": "How To",
+        "slug": "howto",
+        "source": "howto/index.md",
+        "children": [
```
Do we plan to have more how to guides soon? This is probably a good idea but just checking whether there is an immediate need for a 3rd level for this one doc.

dashohoxha replied:
> Do we plan to have more how to guides soon?

Sure. I am planning to add some more. You can also add some if you wish.

> This is probably a good idea but just checking whether there is an immediate need for a 3rd level for this one doc.

I think that the "How To" section is going to become a long one (having lots of items), and eventually it might need to be split from the User Guide as a top-level section on its own. For the time being it is Ok nested under the User Guide.

jorgeorpinel replied:
Sounds good but I think it's best to create the 3rd level when more guides are added. What do you think @shcheklein?

shcheklein replied:
yes, I think we'll have to introduce a how to section soon. There are a few tickets that start with "How to ..." already - probably good candidates for that. May be we can move things like "Running on Windows" into this section>

jorgeorpinel replied:
OK resolving this then. Thanks

shcheklein replied:
@jorgeorpinel this is the discussion around How to that happened a while ago ^^

jorgeorpinel commented on static/docs/user-guide/howto/index.md:
```diff
@@ -0,0 +1,3 @@
+# How To
+
+- [Keep Data and Cache on External Drive](/doc/user-guide/howto/keep-data-and-cache-on-external-drive)
```
See https://github.com/iterative/dvc.org/issues/727#issuecomment-545756811.

dashohoxha replied:
I agree. But for the time being let's keep `index.md`, until the engine is fixed, and remove it later. It should not be difficult to remove it.

jorgeorpinel replied:
Yes. You're right.

jorgeorpinel commented:
Cool. Does this PR include my commits from the other one, abd2446, b3d5c6b, and f345245?

![image](https://user-images.githubusercontent.com/1477535/67512940-795d6280-f65f-11e9-814f-699547ee6a44.png)

Please `git cherry-pick` them if not, they address several issues that were long standing in that PR.

dashohoxha commented:
> Does this PR include my commits

Sure, it does include all the commits. I have also made some other small modifications.

dashohoxha commented:
> I haven't done an extensive check but just looking at the first file in my first commit, lines 40-43 (abd2446#diff-278c4173e3a65d532906f6a613dc7ce0R40-R43) for example are not included in this PR.

I did some regular updates and I found that windows link superfluous and out of place, since this HowTo does not cover at all the windows case, there are Linux commands everywhere (which might also work in Mac, but no guarantee about it). If you want to cover the Windows case maybe you need to add another HowTo which does it thoroughly.

> I guess there was a reason to so it, but having closed the original PR and opened a new one from a different branch makes it extra time consuming to track and review the continuity in these changes. I tried just doing git cherry-pick to apply my work again, but since the file is located somewhere else, this method does not work.

You don't need to track the commits. Just read the HowTo and see whether it is in a good shape or not: https://dvc-org-pr-732.herokuapp.com/doc/user-guide/howto/keep-data-and-cache-on-external-drive

> We'll probably have to manually compare and reapply my changes (combined here: 482e1c7...f345245) vs those in this PR.

If I have removed something and you re-apply your changes (that you don't even remember why they were added in the first place), then you are going to undo my work.
Remember that those changes were meant for a UseCase, and this is a HowTo, and I believe you know the difference, don't you?

In any case, whatever changes/improvements you may think of, I am sure that they are not blocking changes. So let's merge first this PR, and make the changes later during regular updates.

jorgeorpinel commented:
> I did some regular updates and I found that windows link superfluous and out of place...

You may find it out of place but I added that note because Ivan suggested it was necessary. Refer to https://github.com/iterative/dvc.org/pull/565#pullrequestreview-276195808. I think we should discuss and decide, not just discard previous work.

> You don't need to track the commits... you don't even remember why they were added in the first place...

All of my changes had a similar reason, which is why I'm tracking them. It's also how we collaborate. But don't worry, I am going to apply everything again myself when I get a chance (keeping in mind this is now a how-to, not a use-case), and I'll be happy to discuss any review comments you open in this PR on things you think are not needed – you're often right. 

> you are going to undo my work...

I don't see how, but yes, we're gonna have to do some double work unfortunately, given the PR switch. But again, I do understand it helps us have a clean slate, which is nice, since the previous PR was stale and too long. So let's just focus on getting it done.

> they are not blocking changes...

Again, this we all have to decide together.

> p.s. There are some pending review comments by Ivan in the old PR you may want to double check, for example: https://github.com/iterative/dvc.org/pull/565#pullrequestreview-276195595, https://github.com/iterative/dvc.org/pull/565#pullrequestreview-276196319, and https://github.com/iterative/dvc.org/pull/565#pullrequestreview-278196181 for example.

jorgeorpinel commented:
(I'm hiding these couple long comments above ^ to simplify reading this PR.)

shcheklein commented:
@dashohoxha it has really created a problem reviewing this PR. Please, move some of the concerns I and/or @jorgeorpinel raised in the initial one with your responses. Otherwise you force us to do this. I'm not sure why there was a need to create a second one instead of iterating on the first since we effectively lost a lot of context by doing this.

jorgeorpinel commented:
> - This https://github.com/iterative/dvc.org/pull/732#pullrequestreview-307512983 I'll address myself ASAP.

Done! Most of my old changes were indeed already in this new version but I had to review line by line to double check. A few were gone and I added again. I also took the opportunity to make small misc. wording improvements, etc. Feel free to review and comment on this latest version and I'll address.

Glad we're finally getting it over with. Thanks.

shcheklein commented on static/docs/user-guide/howto/keep-data-and-cache-on-external-drive.md:
```diff
@@ -0,0 +1,200 @@
+
+## Create a DVC project with external cache
+
+An [external cache](/doc/user-guide/external-outputs) resides outside of the
```
link is broken

shcheklein commented on static/docs/user-guide/howto/keep-data-and-cache-on-external-drive.md:
```diff
@@ -0,0 +1,200 @@
+
+Sometimes the data may be stored on an
+[external hard drive](https://whatis.techtarget.com/definition/external-hard-drive).
+Usually such data is huge, which means that it won't fit on our local drive, and
```
This motivation is not enough. Do you remember where did this case come from? It's not clear why do I need two drives, what advantages does this configuration has if I store everything - input data, results, etc outside? I understand the case when we do this with S3, for example.

dashohoxha replied:
> Do you remember where did this case come from?

Normally I would not remember, however the issue #563 has some context about it. A user has requested help for this situation on the chat. I am not sure why he has to keep the data on an external drive, but keep the project on his computer. But this is how he prefers to organize his work.

Recently I have seen another discussion where a user wants to keep all the data and cache on an external directory, on a fixed absolute path, so that he can search easily the code of the project. Again I am not sure whether this is the most efficient solution for him, but this is what he does. (Sorry, I cannot find a reference for this discussion.)
I would say that the external-data case is very similar to the external-drive case.

> I understand the case when we do this with S3, for example.

I am not sure about the benefits of using this approach for the S3 case, because the data would still have to be downloaded, in order to be processed locally, and it would still have to be uploaded, in order to be saved remotely. The benefits are not so clear, in my opinion. You might be able to save some local storage space, but this comes at the cost of increased downloads and uploads.

shcheklein commented on static/docs/user-guide/howto/keep-data-and-cache-on-external-drive.md:
````diff
@@ -0,0 +1,200 @@
+```dvc
+$ dvc add /mnt/external-drive/data/raw
+...
+$ dvc run \
````
if you do this, it's better to have a separate directory per project on the external drive. Or, again, I don't quite understand the motivation and use case behind this configuration.

dashohoxha replied:
> if you do this, it's better to have a separate directory per project on the external drive

I don't understand your request here. It is just an example, why should we talk about several projects?

> I don't quite understand the motivation and use case behind this configuration.

I have tried to explain it here: https://github.com/iterative/dvc.org/pull/732#discussion_r339773638

I will try to revise the motivation including this one too:

> Recently I have seen another discussion where a user wants to keep all the data and cache on an external directory, on a fixed absolute path, so that he can search easily the code of the project.

dashohoxha commented:
> The biggest concern is that I don't understand the motivation for this combination.

I left a comment here: https://github.com/iterative/dvc.org/pull/732#discussion_r339773638

dashohoxha commented:
I changed the title and rewrote the motivation. I hope it looks better now.

I was tempted again to remove the reference about Windows (https://github.com/iterative/dvc.org/pull/732/files#diff-e67a9ecb076e21d068e0f5559ed64582R30) but I didn't dare. @jorgeorpinel it really seems out of place to me, please think about removing it again.

shcheklein commented:
btw, @dashohoxha how does this one correlate with the new data sharing section?

dashohoxha commented:
> how does this one correlate with the new data sharing section?

I don't see any correlation.
Maybe there is some correlation with the external data management (it could be an example of it), but I think that it deserves to be a standalone HowTo.

shcheklein commented on static/docs/user-guide/howto/external-data-and-cache.md:
````diff
@@ -0,0 +1,151 @@
+we have to make sure that we can read and write the `/data/` directory:
+
+```dvc
+$ sudo chown <username>: -R /data/
````
I think this is not right to provide chown, chmod and other commands to manage permissions. They are not actionable anyway. Exact commands you need to run depend on the case. Some general suggestions to check the permissions should be enough?

shcheklein commented on static/docs/user-guide/howto/external-data-and-cache.md:
```diff
@@ -0,0 +1,151 @@
+# Keep Data and Cache Outside the Project
+
+Sometimes we would like to keep the data and cache outside the project
```
not sure how is it related to being able to search code tbh, could you elaborate?

shcheklein commented on static/docs/user-guide/howto/external-data-and-cache.md:
```diff
@@ -0,0 +1,151 @@
+Sometimes we would like to keep the data and cache outside the project
+directory, so that we can search easily the code of the project.
+
+Keeping the data outside of the project, on a fixed absolute path, is also
```
it's a a very tricky case - if multiple projects use the same data how do they switch different versions?

shcheklein commented:
I don't see how does it replace #565 or fixes #563. It's a very specific how-to focused on external dependencies and output on a local machine. Also, it reminds me a lot some of the other data managements sections you put together.

shcheklein commented:
closing this as stale
